packages/middleware-change-page-title/tests/index.test.js

```diff
@@ -0,0 +1,183 @@
+    expect(setTitle).not.toHaveBeenCalled()
+    const request = {}
+    const next = jest.fn(() => 'nextvalue')
+    const result = await middleware(request, next)
+    expect(api.getState).toHaveBeenCalledTimes(1)
+    expect(api.getState).toHaveBeenCalledWith()
+    expect(createSelector).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledWith('test state')
+    expect(setTitle).not.toHaveBeenCalled()
+    expect(next).toHaveBeenCalledTimes(1)
+    expect(result).toEqual('nextvalue')
+  })
+
+  test('Does not set the title if it is undefined', async () => {
+    const setTitle = jest.fn()
+    const selectTitle = jest.fn(() => undefined)
+    const createSelector = jest.fn(() => selectTitle)
+    const creator = changePageTitle({
+      isServer: () => false,
+      setTitle,
+      createSelector,
+    })
+    const api = {
+      getState: jest.fn(() => 'test state'),
+    }
+    const middleware = creator(api)
+    expect(createSelector).toHaveBeenCalledTimes(1)
+    expect(createSelector).toHaveBeenCalledWith('title', undefined)
+    expect(api.getState).not.toHaveBeenCalled()
+    expect(selectTitle).not.toHaveBeenCalled()
+    expect(setTitle).not.toHaveBeenCalled()
+    const request = {}
+    const next = jest.fn(() => 'nextvalue')
+    const result = await middleware(request, next)
+    expect(api.getState).toHaveBeenCalledTimes(1)
+    expect(api.getState).toHaveBeenCalledWith()
+    expect(createSelector).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledWith('test state')
+    expect(setTitle).not.toHaveBeenCalled()
+    expect(next).toHaveBeenCalledTimes(1)
+    expect(result).toEqual('nextvalue')
+  })
+
+  test('Does not set the title if it is a function', async () => {
+    const setTitle = jest.fn()
+    const selectTitle = jest.fn(() => () => {})
+    const createSelector = jest.fn(() => selectTitle)
+    const creator = changePageTitle({
+      isServer: () => false,
+      setTitle,
+      createSelector,
+    })
+    const api = {
+      getState: jest.fn(() => 'test state'),
+    }
+    const middleware = creator(api)
+    expect(createSelector).toHaveBeenCalledTimes(1)
+    expect(createSelector).toHaveBeenCalledWith('title', undefined)
+    expect(api.getState).not.toHaveBeenCalled()
+    expect(selectTitle).not.toHaveBeenCalled()
+    expect(setTitle).not.toHaveBeenCalled()
+    const request = {}
+    const next = jest.fn(() => 'nextvalue')
+    const result = await middleware(request, next)
+    expect(api.getState).toHaveBeenCalledTimes(1)
+    expect(api.getState).toHaveBeenCalledWith()
+    expect(createSelector).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledTimes(1)
+    expect(selectTitle).toHaveBeenCalledWith('test state')
+    expect(setTitle).not.toHaveBeenCalled()
+    expect(next).toHaveBeenCalledTimes(1)
+    expect(result).toEqual('nextvalue')
+  })
```

Comment:
good lord, these tests are cray, but it looks like you res a lot of the api's current state

perhaps I'm being foolish, but James wanted to create this automates jest test generators(using redux dev tools) to generate all the tests as you use the app.

If you have the time, id love a quick note about how you were able to create so, so many tests without an ungodly amount of enzyme or jest logic in there. There's more `expect`s than really anything else. That's not what you see in unit tests. 

If you did what it looks like, then i am thoroughly impressed and would like to understand a little more (i might be enlightened as i read down the rest of the PR

Reply:
It's not so much that it tests the api's state, rather it tests that the subject of the test (the middleware) does the right things to its `api` argument (which is actually a mocked out version of the real api).

There is a sliding scale from isolated tests of one module (unit tests) to tests that touch many modules in a complex system (integration tests). Generally speaking, with integration tests much less testing code is necessary to get to say 80% coverage, but they come with disadvantages:
- When the tests fail, its often difficult to tell which part of the code is wrong
- The tests often are slow to run, because they have to run lots of code (tests * code size) and likely do some sort of IO (requests, etc). If you change a single file, you probably have to run all/most of your tests again, because all the tests touch all the files.
- Often you have to do extra work to the testing environment to accommodate all the assumptions in the system (e.g. making sure `window` is defined or not, monkey patching random number generators, etc)
- There might be particular behaviours in certain modules that only happen in very specific sets of circumstances - its virtually impossible to find all of these by using the entire system from the outside. Getting to 80% coverage is easy, but getting to 100% is basically impossible.

Unit tests on the other hand require one test for each module you want to test, so there's a whole lot more initial work to get to 80% test coverage, but there are lots of advantages:
- Getting to 100% test coverage is just 20% more work - you don't have to guess which parts of your code will get used when you use the entire system together - you can go through each file and make sure the tests hit all of the branches in each of them.
- When a test fails, it points somewhere very close to where the error is, because its testing a small thing. Finding the error is usually easy.
- The tests run fast, because each test only executes one module, and doesn't do any IO or have any external dependencies. It's easy to run a small subset of tests when you change a file, because only some of the tests test that file.
- It encourages good design, because your code needs to be in small isolated modules in order to be unit tested (otherwise you end up needing to mock lots of things out and its hard to determine exactly what each module is meant to do)

Reply:
Automatically generated tests could be useful - I'm not really familiar with any tools but I can kind of imagine that it could work. JS is a hard language to do it in though - so much global state everywhere.

Reply:
This is what changed my mind and made me understand the purpose of unit tests - its a great talk: https://www.youtube.com/watch?v=VDfX44fZoMc

Also interesting (TL;DR: its even better and easier to have pure functions as much as possible and not have to mock anything out): https://www.youtube.com/watch?v=yTkzNHF6rMs

I think actually Rudy/RFR is a great tool for making an app unit testable. Imagine a react/redux/rfr-rudy app where:
- Every react component is a pure function (`App = f(state)`)
- Most of the applications logic is in reducers (pure functions `state = f(state, action)`),
- Maybe you have selectors (also pure functions of the redux state)
- There is maybe a small "imperative shell" centered around RFR/Rudy which handles side effects, asynchronicity, and other stuff that is hard to put in pure functions.

This design is very easy to write unit tests for. The components and reducers are all pure functions, so testing them is really simple - you call them with a selection of arguments, and check that they return the right thing. You don't need to set up the entire redux state of the app when it runs, just the part relevant to each selector/reducer/component. You don't need to mock out an API client or the `window` object, because they aren't used. Its a bit harder for the thunks, etc. You might have to mock out an API client, and you might have to simulate a few steps to test all the behaviour (e.g. make one API call succeed, and then the next one fail). But at least the behaviour is simple, and there isn't much of that non pure functional code).

You can unit test almost everything and probably have a super reliable/maintainable app.

Imagine if instead "everything is a component". Components all have state. Maybe they also dispatch actions which change the redux state and then do stuff in `getDerivedStateFromProps` after other thunks outside the component run. Maybe that again depends on the component's state. Maybe some components change the page title, do API requests, or other global stuff.

You can't easily unit test that. Even testing any stateful component is way more effort - there's so many more options and so much more work to get the component to do all the things it does. If the component also touches other stuff, you might have to mock out the global `window` object, or more of the DOM, or an API client, or some other global thing. Its alot of effort, and even then, the interface between that component and the rest of the system is so complicated its almost impossible to write tests that convince you the whole system works. How can you answer the question "Is this component correct" if its hard to even say what its meant to do with all those global variables and when?

Reply:
Lol that’s James philosophy. Everything is app(state) make scale manageable

Reply:
So I think Remixx comes into play or something. But the basis was that everything was proxied and redux knew every react action because it managed it all. Anton found a way to proxy stuff at a performant level. So be React doing anything. We have a redux like bhistory and payload. Playing back the outcome of the business logic, and replaying the actions that got it there is the equivalent of a unit test to me.  It automated my proxing react

Reply:
Really education response on tests. I’m. Or very good at them and learned. Copied it to a note

Reply:
Actually I can kind of imagine that - because the remixx stuff has instrumentation, it mght be possible to automatically construct a subset of the redux state for the purposes of unit tests

Reply:
I languished in darkness and confusion for years until I watched that video - could not recoommend more highly

Reply:
Oh I will watch them. I hate tests and writing them. With Rudy, we can snapshot the whole api and state. Along with each action performed. Remixx proxies all internal events via redux allowing us to track lifecycle stuff (I think)